### PR TITLE
Memory optimisation in matmul

### PR DESCRIPTION
**Pull request by YijieYin**
Wrote together with @MonEstCha. 

The idea is to only move chunked rows/columns into GPU, to save GPU memory. For now we assume there is sufficient CPU memory, since that's typically the case on high performance computing clusters. 

Key part of the function is as follows: 
```Python
for i in range(step_number):
  if i == 0: # keep it as is 
      out_tensor = inprop_tensor.clone()
  else:
      out_tensor_new = torch.full(
          (size, size), 0.0, dtype=torch.float32) # initialise a full matrix in CPU 
      colLow = 0
      colHigh = chunkSize
      for colChunk in range(chunks):  # iterate chunks colwise
          rowLow = 0
          rowHigh = chunkSize

          in_col = inprop_tensor[:, colLow:colHigh].to(
              device)  # shape: size x chunkSize; on GPU

          for rowChunk in range(chunks):  # iterate chunks rowwise
              in_rows = out_tensor[rowLow:rowHigh, :].to(
                  device)  # shape: chunkSize x size; on GPU
              # fill up the initialised matrix chunk by chunk 
              out_tensor_new[rowLow:rowHigh, colLow:colHigh] = torch.matmul(
                  in_rows, in_col).to('cpu')  # shape: chunkSize x chunkSize; on CPU

              rowLow += chunkSize
              rowHigh += chunkSize
              rowHigh = min(rowHigh, size)
          colLow += chunkSize
          colHigh += chunkSize
          colHigh = min(colHigh, size)
          
          # use the result for the new iteration 
          out_tensor = out_tensor_new 
         # then output the thresholded version 
```

And using `chunkSize` instead of `n_chunks` as input, since that's more directly related to GPU memory calculations, and is independent from the input matrix size. 

In the future, would be nice to write/read chunks into CPU memory, so it it runs in theory on a laptop that ideally has a GPU (though this would take long).

### CODE REVIEW

File: connectome_interpreter/compress_paths.py [hunk 1]
```diff
@@ -48,25 +49,78 @@ def compress_paths(inprop, step_number, threshold=0, output_threshold=1e-4, root
...
+                        out_tensor_new[rowLow:rowHigh, colLow:colHigh] = torch.matmul(
+                            in_rows, in_col).to('cpu')  # shape: chunkSize x chunkSize; on CPU
+                        # print('after matmul')
+                        # print(torch.cuda.max_memory_allocated() / 1024**2)
+
+                        rowLow += chunkSize
+                        rowHigh += chunkSize
+                        rowHigh = min(rowHigh, size)
+
+                        del in_rows
```
Comment (jftsang) [inline connectome_interpreter/compress_paths.py:103]: This doesn't actually free any memory. `del` deletes the name but by itself it doesn't necessarily free the underlying object (which was created on line 90). 

If your `del` statement removes the last reference to that object, then that object is marked for garbage collection (see [this thread](https://stackoverflow.com/questions/1316767/how-can-i-explicitly-free-memory-in-python)). But in your case that is going to happen on the next iteration anyway, when you assign the name `in_rows` to a different object.

```suggestion
```

(The only exception is at the very end of the loop, between when this loop terminates and when the function returns, when the name falls out of scope. So you _could_ put your `del` statements there, but (a) unless you're doing something memory-intense at the end of the function, that doesn't really give you any benefit; and (b) you might need to deal with the annoying edge case where you have zero iterations in your loop and therefore never assign the name in the first place.)

File: connectome_interpreter/compress_paths.py [hunk 2]
```diff
@@ -48,25 +49,78 @@ def compress_paths(inprop, step_number, threshold=0, output_threshold=1e-4, root
...
+                    rowLow = 0
+                    rowHigh = chunkSize
+
+                    in_col = inprop_tensor[:, colLow:colHigh].to(
+                        device)  # shape: size x chunkSize; on GPU
+                    # print('after setting in_col')
+                    # print(torch.cuda.max_memory_allocated() / 1024**2)
+
+                    for rowChunk in range(chunks):  # iterate chunks rowwise
+                        torch.cuda.reset_peak_memory_stats()
```
Comment (jftsang) [inline connectome_interpreter/compress_paths.py:89]: If this is for stats and analytics, it would make a bit more sense to move this after the next assignment. See the note on `del` later.

File: connectome_interpreter/compress_paths.py [hunk 3]
```diff
@@ -48,25 +49,78 @@ def compress_paths(inprop, step_number, threshold=0, output_threshold=1e-4, root
...
+                        # print(torch.cuda.max_memory_allocated() / 1024**2)
+
+                        rowLow += chunkSize
+                        rowHigh += chunkSize
+                        rowHigh = min(rowHigh, size)
+
+                        del in_rows
+                        # print(
+                        #     f'max memory in {i}: {torch.cuda.max_memory_allocated()/ 1024**2:.2f} MB')
+                    del in_col
```
Comment (jftsang) [inline connectome_interpreter/compress_paths.py:106]: ditto

File: connectome_interpreter/compress_paths.py [hunk 4]
```diff
@@ -48,25 +49,78 @@ def compress_paths(inprop, step_number, threshold=0, output_threshold=1e-4, root
...
+                    colLow += chunkSize
+                    colHigh += chunkSize
+                    colHigh = min(colHigh, size)
+
+                out_tensor = out_tensor_new
+                del out_tensor_new
```
Comment (jftsang) [inline connectome_interpreter/compress_paths.py:112]: ditto

File: connectome_interpreter/compress_paths.py [hunk 5]
```diff
@@ -48,25 +49,78 @@ def compress_paths(inprop, step_number, threshold=0, output_threshold=1e-4, root
...
+                # Clear PyTorch CUDA cache
+                torch.cuda.empty_cache()
```
Comment (jftsang) [inline connectome_interpreter/compress_paths.py:114]: you probably do need this though, I'm not sure

File: connectome_interpreter/compress_paths.py [hunk 6]
```diff
@@ -78,11 +132,86 @@ def compress_paths(inprop, step_number, threshold=0, output_threshold=1e-4, root
                 out_csc.eliminate_zeros()
 
             steps_fast.append(out_csc)
-            torch.cuda.empty_cache()
+
+    # remove all variables
+    del out_tensor, inprop_tensor
```
Comment (jftsang) [inline connectome_interpreter/compress_paths.py:137]: ditto
```suggestion
```

File: connectome_interpreter/compress_paths.py [hunk 7]
```diff
@@ -78,11 +132,86 @@ def compress_paths(inprop, step_number, threshold=0, output_threshold=1e-4, root
...
+    torch.cuda.empty_cache()
 
     return steps_fast
 
 
+# # below: not chunked version
+# def compress_paths(inprop, step_number, threshold=0, output_threshold=1e-4, root=False):
```
Comment (jftsang) [inline connectome_interpreter/compress_paths.py:144]: Uncomment and rename. Commented-out code tends to rot, because your linters and type checkers will ignore it.

If this function is meant to be _mostly_ the same as the preceding function, then you could phrase the docstring as something like:
```python
def compress_paths_without_chunking(...):
    """As above, but without chunking, and instead...
    """
```
highlighting the difference.

File: tests/test_compress_paths.py [hunk 8]
```diff
@@ -2,11 +2,14 @@
...
 # Performance: If necessary, tests that evaluate the function’s performance to ensure it meets required standards.
+from connectome_interpreter.compress_paths import compress_paths, compress_paths_memeff
 import unittest
 import numpy as np
 from scipy.sparse import csr_matrix
 # Adjust the import according to your package structure
-from connectome_interpreter.compress_paths import compress_paths
+import sys
+# setting path
+sys.path.append('../connectome_interpreter')
```
Comment (jftsang) [inline tests/test_compress_paths.py:12]: This only works if your working directory, when you invoke `pytest`, is `tests`. 
```suggestion
from pathlib import Path  # but, alphabetise your imports
sys.path.append(Path(__file__).parent.parent / 'connectome_interpreter')
```

I'm guessing you need this because you haven't `pip install connectome_interpreter`'d, or you want to force Python to use the local version rather than an installed version?

File: tests/test_compress_paths.py [hunk 9]
```diff
@@ -2,11 +2,14 @@
 # Boundary Conditions: Tests that challenge the edges of input domains and operational boundaries.
 # Error Conditions: Tests that ensure the function handles errors gracefully, such as bad input values.
 # Performance: If necessary, tests that evaluate the function’s performance to ensure it meets required standards.
+from connectome_interpreter.compress_paths import compress_paths, compress_paths_memeff
```
Comment (jftsang) [inline tests/test_compress_paths.py:5]: Standard library imports first, then third-party packages, then your own code

**Review by jftsang — CHANGES_REQUESTED** [review]

**Review by jftsang — CHANGES_REQUESTED** [review]
There's also an `ImportError`, see the checks

File: connectome_interpreter/compress_paths.py [hunk 10]
```diff
@@ -48,25 +49,78 @@ def compress_paths(inprop, step_number, threshold=0, output_threshold=1e-4, root
...
+    #     f'max memory before start: {torch.cuda.max_memory_allocated()/ 1024**2:.2f} MB')
 
     with torch.no_grad():
         for i in tqdm(range(step_number)):
+
             if i == 0:
                 out_tensor = inprop_tensor.clone()
             else:
-                out_tensor = torch.matmul(out_tensor, inprop_tensor)
+                # print(f'begin iteration {i}')
```
[the remark this answers is not in the document]
Comment (YijieYin) [reply connectome_interpreter/compress_paths.py:70]: Thanks! But after getting the out_tensor regardless of i, it's thresholded (and optionally i-rooted) and turned into a sparse matrix. Everything being indented with one less tab is nice, but the thresholding and output would be skipped.

File: tests/test_compress_paths.py (hunk 8 above)
Comment (YijieYin) [reply tests/test_compress_paths.py:12]: I think this came from @MonEstCha.

File: connectome_interpreter/compress_paths.py (hunk 2 above)
Comment (YijieYin) [reply connectome_interpreter/compress_paths.py:89]: Sorry just forgot to remove it. Removed now.

File: connectome_interpreter/compress_paths.py (hunk 1 above)
Comment (YijieYin) [reply connectome_interpreter/compress_paths.py:105]: Thanks for your comment. I actually looked into it with [this colab notebook](https://colab.research.google.com/drive/1nq2wGLPn_Bip9w1BMqqlAoOQyxpqEKqb?usp=sharing), and then dragging the downloaded files into [Pytorch's memory visualisation tool](https://pytorch.org/memory_viz). Strangely, removing `in_cols` and `in_rows` seem to reduce the max memory used by a little, but not the other ones: 
This one has all variables removed (max around 1.8GB): 
image

And this one keeps `in_cols` and `in_rows` (max a bit over 1.8GB): 
image

However keeping all variables doesn't seem to make a difference: 
image

Please lmk if you think this could be some artefact though.

File: tests/test_compress_paths.py (hunk 8 above)
Comment (MonEstCha) [reply tests/test_compress_paths.py]: Oh, my bad. I somehow couldn't get it running a different way on my local machine. Please ignore this change.
